### PR TITLE
Query : Get Baker List

### PR DESCRIPTION
**Pull request by parv0888**
## Purpose
Added support for query `GetBakerList`

## Changes
Added support for query `GetBakerList`

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance
By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java [hunk 1]
```diff
@@ -303,6 +304,28 @@ public PeerStatistics getPeerStatistics(final boolean includeBootstrappers) {
...
+     * Get the IDs of the bakers registered in the given block.
+     * Note that this list is in general different from the bakers that are
+     * returned as part of [get_birk_parameters](Client::get_birk_parameters).
+     * The latter are the bakers that are eligible for baking at the time of
+     * the block.
+     * @param blockHash {@link Hash} of the block bakers are to be retrieved.
+     * @return Parsed {@link ImmutableList} of {@link BakerId}
+     * @throws Exception When the returned JSON is null.
+     */
+    public ImmutableList<BakerId> getBakerList(Hash blockHash) throws Exception {
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/Client.java:317]: I think we should use the `BlockNotFoundException` in these situations: https://github.com/Concordium/concordium-java-sdk/blob/main/concordium-sdk/src/main/java/com/concordium/sdk/exceptions/BlockNotFoundException.java

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import com.concordium.sdk.serializing.JsonMapper;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import concordium.ConcordiumP2PRpc;
+import lombok.Data;
+import lombok.Getter;
+import lombok.ToString;
+
+/**
+ * Short id of the baker.
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java:13]: I'm not sure what 'short' means here.
It is simply the account index of the account baking i.e. the account registered as a baker.

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java [hunk 3]
```diff
@@ -303,6 +304,28 @@ public PeerStatistics getPeerStatistics(final boolean includeBootstrappers) {
...
+        val res = server()
+                .getBakerList(ConcordiumP2PRpc.BlockHash.newBuilder().setBlockHash(blockHash.asHex()).build());
+
+        BakerId[] bakerIdArray = BakerId.fromJsonArray(res);
+        if (Objects.isNull(bakerIdArray)) {
+            throw new Exception(String.format("Could not retrieve Baker Ids at block %s", blockHash.asHex()));
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/Client.java:323]:
```suggestion
            throw new Exception(String.format("Could not retrieve baker ids for block %s", blockHash.asHex()));
```
Also please use `BlockNotFoundException ` as mentioned above.

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java [hunk 4]
```diff
@@ -303,6 +304,28 @@ public PeerStatistics getPeerStatistics(final boolean includeBootstrappers) {
         return PeerStatistics.parse(value);
     }
 
+    /**
+     * Get the IDs of the bakers registered in the given block.
+     * Note that this list is in general different from the bakers that are
+     * returned as part of [get_birk_parameters](Client::get_birk_parameters).
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/Client.java:310]: Please convert to proper javadoc. Also 'get birk parameters' is not currently in the branch?

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good. Minor comments 👍

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java (hunk 2 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java:13]: changed. done

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java (hunk 1 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/Client.java:317]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java (hunk 3 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/Client.java:323]: `BlockNotFoundException` currently does not have a string constructor argument. Should we add one?

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java (hunk 4 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/Client.java:310]: done

File: README.md [hunk 5]
```diff
@@ -373,6 +379,12 @@ val peers = client.getPeerList(true);
 SemVer version = client.getVersion();
 ```
 
+#### getBakerList
+```java
+val bakerList = client
```
Comment (EmilLa1) [inline README.md:384]: Please specify types in the documentation.

```suggestion
ImmutableList<BakerId> bakerList = client
```

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    BakerId(long index) {
+        this.id = AccountIndex.from(index);
+    }
+
+    @Override
+    public String toString() {
+        return this.id.toString();
+    }
+
+    public static BakerId[] fromJsonArray(ConcordiumP2PRpc.JsonResponse res) {
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java:32]: Can you please provide a simple unit test with a mocked JSON response?

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good, minor comments.

File: README.md (hunk 5 above)
Comment (parv0888) [reply README.md:384]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java (hunk 6 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/bakerlist/BakerId.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java (hunk 3 above)
Comment (EmilLa1) [reply concordium-sdk/src/main/java/com/concordium/sdk/Client.java]: It takes a `Hash` which I believe exactly is what `blockHash` is? So if you just omit the `.asHex()` I think you should be able to use it.

Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/Client.java]: Done, This has been refactored to just use block hash. Please do check

**Review by EmilLa1 — APPROVED** [review]
Looks good.

**Pull request merged**